can/interfaces/canalystii.py

```diff
@@ -69,132 +58,494 @@ class VCI_CAN_OBJ(Structure):
-                logger.error("VCI_StartCAN Error")
+        elif bitrate is not None:
+            if bitrate in TIMING_DICT:
+                Timing0, Timing1 = TIMING_DICT[bitrate]
+            else:
                 self.shutdown()
-                return
+                raise CanalystIIBitrateNotSupported(errno=bitrate)
+
+        elif None in (Timing0, Timing1):
+            self.shutdown()
+            raise CanalystIIInvalidTimings()
+
+        init_config.Timing0 = Timing0
+        init_config.Timing1 = Timing1
+
+        self.init_config = init_config
+
+        status = CANalystII.VCI_InitCAN(
+            VCI_USBCAN2,
+            device,
+            channel,
+            byref(self.init_config)
+        )
+
+        if status != STATUS_OK:
+            self.shutdown()
+            raise CANalystIIStartupError(
+                "VCI_InitCAN Error, incorrect channel number?",
+                status
+            )
+
+        status = CANalystII.VCI_StartCAN(
+            VCI_USBCAN2,
+            device,
+            channel
+        )
+
+        if status != STATUS_OK:
+            logger.error(
+                "VCI_StartCAN Error ({0}), device:{1} "
+                "channel:{2}".format(status, device, channel)
+            )
+            logger.info('receiving failed to start, only able to transmit.')
+
+    def board_info(self) -> Dict:
+        """
+        Adapter Information
+
+        :return: dictionary containing the following keys
+
+            * `hardware_version`: `int`
+            * `firmware_version`: `int`
+            * `driver_version`: `int`
+            * `library_version`: `int`
+            * `irq`: `int`
+            * `num_can_interfaces`: `int`
+            * `serial_number`: `str`
+            * `hardware_type`: `str`
+
+        :rtype: dict
+        """
+        return {key: value for key, value in self._b_info.items()}
 
-    def send(self, msg, timeout=None):
+    def send(self, msg: Message, timeout=None):
         """
+        Send CAN Frame
 
         :param msg: message to send
+        :type msg: :class:`can.Message`
+
         :param timeout: timeout is not used here
-        :return:
+        :type timeout: None
+
+        :return: `True` if frame was sent else `False`
+        :rtype: bool
         """
-        extern_flag = 1 if msg.is_extended_id else 0
-        raw_message = VCI_CAN_OBJ(
-            msg.arbitration_id,
-            0,
-            0,
-            1,
-            msg.is_remote_frame,
-            extern_flag,
-            msg.dlc,
-            (c_ubyte * 8)(*msg.data),
-            (c_byte * 3)(*[0, 0, 0]),
-        )
 
-        if msg.channel is not None:
-            channel = msg.channel
-        elif len(self.channels) == 1:
-            channel = self.channels[0]
+        can_objs = (VCI_CAN_OBJ * 1)()
+
+        arbitration_id = msg.arbitration_id
+        if arbitration_id & 0x80000000:
```

Comment:
Only `msg.is_extended_id` is used.

Reply:
my point to this is an unknowing user could place say an arbitration id of `0x9003E099` and not set the is_extended flag, with no checks in place to see if the arbitration id is a 32 bit id then it can cause problems for the user making it difficult to nail down the actual issue.

`arbitration_id = 0x9003E099 & 0x7FFFFFFF  # 0x1003E099`
`is_extended = bool(0x9003E099 & 0x80000000)`


a small modification of the Message class would eliminate any possibility of having an arbitration is set that is an extended id. and you can dd a property to get the full arbitration id including the set bit for extended.

```python

class Message:
    """
    The :class:`~can.Message` object is used to represent CAN messages for
    sending, receiving and other purposes like converting between different
    logging formats.

    Messages can use extended identifiers, be remote or error frames, contain
    data and may be associated to a channel.

    Messages are always compared by identity and never by value, because that
    may introduce unexpected behaviour. See also :meth:`~can.Message.equals`.

    :func:`~copy.copy`/:func:`~copy.deepcopy` is supported as well.

    Messages do not support "dynamic" attributes, meaning any others than the
    documented ones, since it uses :attr:`~object.__slots__`.
    """

    __slots__ = (
        "timestamp",
        "arbitration_id",
        "is_extended_id",
        "is_remote_frame",
        "is_error_frame",
        "channel",
        "dlc",
        "data",
        "is_fd",
        "is_rx",
        "bitrate_switch",
        "error_state_indicator",
        "__weakref__",  # support weak references to messages
    )

    def __init__(
        self,
        timestamp: float = 0.0,
        arbitration_id: int = 0,
        is_extended_id: bool = True,
        is_remote_frame: bool = False,
        is_error_frame: bool = False,
        channel: Optional[typechecking.Channel] = None,
        dlc: Optional[int] = None,
        data: Optional[typechecking.CanData] = None,
        is_fd: bool = False,
        is_rx: bool = True,
        bitrate_switch: bool = False,
        error_state_indicator: bool = False,
        check: bool = False,
    ):
        """
        To create a message object, simply provide any of the below attributes
        together with additional parameters as keyword arguments to the constructor.

        :param check: By default, the constructor of this class does not strictly check the input.
                      Thus, the caller must prevent the creation of invalid messages or
                      set this parameter to `True`, to raise an Error on invalid inputs.
                      Possible problems include the `dlc` field not matching the length of `data`
                      or creating a message with both `is_remote_frame` and `is_error_frame` set to `True`.

        :raises ValueError: iff `check` is set to `True` and one or more arguments were invalid
        """
        self.timestamp = timestamp

         if arbitration_id & 0x80000000:
             self.is_extended_id = True
             arbitration_id = arbitration_id  & 0x7FFFFFFF 
         else:
             self.is_extended_id = is_extended_id

        self.arbitration_id = arbitration_id
        self.is_remote_frame = is_remote_frame
        self.is_error_frame = is_error_frame
        self.channel = channel
        self.is_fd = is_fd
        self.is_rx = is_rx
        self.bitrate_switch = bitrate_switch
        self.error_state_indicator = error_state_indicator

        if data is None or is_remote_frame:
            self.data = bytearray()
        elif isinstance(data, bytearray):
            self.data = data
        else:
            try:
                self.data = bytearray(data)
            except TypeError:
                err = "Couldn't create message from {} ({})".format(data, type(data))
                raise TypeError(err)

        if dlc is None:
            self.dlc = len(self.data)
        else:
            self.dlc = dlc

        if check:
            self._check()

    @property
    def full_arbitration_id(self):
        if self.is_extended:
            return self.arbitration_id | 0x80000000
        return self.arbitration_id

```

Reply:
I believe that identifying an extended message using a specific bit in the CAN ID is implementation specific and not part of the standard. Or is there some part of the standard that I've missed?

Reply:
you must have missed it.


![image](https://user-images.githubusercontent.com/10932297/99890890-2e769b80-2c21-11eb-98a1-1be4622f6601.png)


Standard Frame CAN 2.0A
![image](https://user-images.githubusercontent.com/10932297/99890972-1bb09680-2c22-11eb-9dd7-1b5dab6b79f5.png)


Field name | Length (bits) | Purpose
-- | -- | --
Start-of-frame | 1 | Denotes the start of frame transmission
11 bit Identifier | 11 | A (unique) identifier which also represents the message priority
Remote transmission request (RTR)| 1 | Must be dominant (0) for data frames and recessive (1) for remote request frames 
Identifier extension bit (IDE) | 1 | Must be dominant (0) for base frame format with 11-bit identifiers
Reserved bit (r0) | 1 | Reserved bit.  Must be dominant (0), but accepted as either dominant or recessive.
Data length code (DLC) | 4 | Number of bytes of data (0–8 bytes)[a]
Data field| 0–64 (0-8 bytes) | Data to be transmitted (length in bytes dictated by DLC field)
CRC | 15 | Cyclic redundancy check
CRC delimiter | 1 | Must be recessive (1)
ACK slot | 1 | Transmitter sends recessive (1) and any receiver can assert a dominant (0)
ACK delimiter | 1 | Must be recessive (1)
End-of-frame (EOF) | 7 | Must be recessive (1)


Extended Frame CAN 2.0B
![image](https://user-images.githubusercontent.com/10932297/99890969-12bfc500-2c22-11eb-9c6a-1720455eb259.png)

Field name | Length (bits) | Purpose
-- | -- | --
Start-of-frame | 1 | Denotes the start of frame transmission
11 bit Identifier | 11 | First part of the (unique) identifier which also represents the message priority
Substitute remote request (SRR) | 1 | Must be recessive (1)
Identifier extension bit (IDE) | 1 | Must be recessive (1) for extended frame format with 29-bit identifiers
18 bit Identifier | 18 | Second part of the (unique) identifier which also represents the message priority
Remote transmission request (RTR) | 1 | Must be dominant (0) for data frames and recessive (1) for remote request frames
Reserved bits (r1, r0) | 2 | Reserved bits which must be set dominant (0), but accepted as either dominant or recessive
Data length code (DLC) | 4 | Number of bytes of data (0–8 bytes)[a]
Data field | 0–64 (0-8 bytes) | Data to be transmitted (length dictated by DLC field)
CRC | 15 | Cyclic redundancy check
CRC delimiter | 1 | Must be recessive (1)
ACK slot | 1 | Transmitter sends recessive (1) and any receiver can assert a dominant (0)
ACK delimiter | 1 | Must be recessive (1)
End-of-frame (EOF) | 7 | Must be recessive (1)


we are interested in "Identifier extension bit (IDE)" which is the bit that gets toggled to on if the frame is an extended frame.
you also have the message priority that is embedded in the 11 bit Identifier.

Which I made a boo boo when I did the last commit for this PR. I added to many F''s and 0's to the checking for an extended frame.

Reply:
I just see that the IDE bit is completely separate from the identifier. Why bit 31? What do the other two bits map to? In case of an 11-bit ID, where are the rest of the bits?

Aren't you just referring to details in the socketcan or Vector implementations?

Reply:
It doesn't really matter i guess, because of how dbc files set the extended bit is different from what the can specification is and can arbitration ids located on the internet for dealing with vehicles being posted different ways and dbc files made for vehicles being made incorrectly there is no easy way to know what is correct.

Reply:
That's just an internal representation in dbc files (and other Vector specific formats and APIs) but their user interfaces always display the correct ID without the extra highest bits. So we shouldn't expose these non-standard details in our API either.